File: class/class6.1/message_dispatcher.cpp
#include <iostream>
#include <vector>
#include <string>
#include <functional>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <random>

struct GreenMessage {
    std::string text;
    int counter;
};

struct BlueMessage {
    double value1;
    double value2;
};

struct OrangeMessage {
    std::string text1;
    std::string text2;
    int integer;
    double value;
};

class MessageDispatcher {
public:
    void publishGreenMessage(const GreenMessage& msg) {
        std::unique_lock<std::mutex> lock(mtx);
        for (auto& subscriber : greenSubscribers) {
            subscriber(msg);
        }
    }

    void publishBlueMessage(const BlueMessage& msg) {
        std::unique_lock<std::mutex> lock(mtx);
        for (auto& subscriber : blueSubscribers) {
            subscriber(msg);
        }
    }

    void publishOrangeMessage(const OrangeMessage& msg) {
        std::unique_lock<std::mutex> lock(mtx);
        for (auto& subscriber : orangeSubscribers) {
            subscriber(msg);
        }
    }

    void subscribeToGreenMessage(std::function<void(const GreenMessage&)> callback) {
        std::unique_lock<std::mutex> lock(mtx);
        greenSubscribers.push_back(callback);
    }

    void subscribeToBlueMessage(std::function<void(const BlueMessage&)> callback) {
        std::unique_lock<std::mutex> lock(mtx);
        blueSubscribers.push_back(callback);
    }

    void subscribeToOrangeMessage(std::function<void(const OrangeMessage&)> callback) {
        std::unique_lock<std::mutex> lock(mtx);
        orangeSubscribers.push_back(callback);
    }

private:
    std::vector<std::function<void(const GreenMessage&)>> greenSubscribers;
    std::vector<std::function<void(const BlueMessage&)>> blueSubscribers;
    std::vector<std::function<void(const OrangeMessage&)>> orangeSubscribers;

    std::mutex mtx;
};

void generateGreenMessages(MessageDispatcher& dispatcher) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dist(1, 100);

    for (int i = 0; i < 10; ++i) {
        GreenMessage msg = {"GreenMessage_" + std::to_string(i), dist(gen)};
        dispatcher.publishGreenMessage(msg);
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}

void generateBlueMessages(MessageDispatcher& dispatcher) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dist(0.0, 100.0);

    for (int i = 0; i < 10; ++i) {
        BlueMessage msg = {dist(gen), dist(gen)};
        dispatcher.publishBlueMessage(msg);
        std::this_thread::sleep_for(std::chrono::milliseconds(150));
    }
}

void generateOrangeMessages(MessageDispatcher& dispatcher) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> intDist(1, 100);
    std::uniform_real_distribution<> doubleDist(0.0, 100.0);

    for (int i = 0; i < 10; ++i) {
        OrangeMessage msg = {
            "OrangeMessage1_" + std::to_string(i),
            "OrangeMessage2_" + std::to_string(i),
            intDist(gen),
            doubleDist(gen)
        };
        dispatcher.publishOrangeMessage(msg);
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
    }
}

int main() {
    MessageDispatcher dispatcher;

    dispatcher.subscribeToGreenMessage([](const GreenMessage& msg) {
        std::cout << "Received GreenMessage: " << msg.text << ", " << msg.counter << std::endl;
    });
    dispatcher.subscribeToBlueMessage([](const BlueMessage& msg) {
        std::cout << "Received BlueMessage: " << msg.value1 << ", " << msg.value2 << std::endl;
    });
    dispatcher.subscribeToOrangeMessage([](const OrangeMessage& msg) {
        std::cout << "Received OrangeMessage: " << msg.text1 << ", " << msg.text2 << ", " << msg.integer << ", " << msg.value << std::endl;
    });

    std::thread greenThread(generateGreenMessages, std::ref(dispatcher));
    std::thread blueThread(generateBlueMessages, std::ref(dispatcher));
    std::thread orangeThread(generateOrangeMessages, std::ref(dispatcher));

    greenThread.join();
    blueThread.join();
    orangeThread.join();

    return 0;
}

File: home/home6.1/message_dispatcher.swift
import Foundation

struct GreenMessage {
    let text: String
    let counter: Int
}

struct BlueMessage {
    let value1: Double
    let value2: Double
}

struct OrangeMessage {
    let text1: String
    let text2: String
    let integer: Int
    let value: Double
}

class MessageDispatcher {
    private var greenSubscribers: [UUID: (GreenMessage) -> Void] = [:]
    private var blueSubscribers: [UUID: (BlueMessage) -> Void] = [:]
    private var orangeSubscribers: [UUID: (OrangeMessage) -> Void] = [:]
    
    private let queue = DispatchQueue(label: "com.message.dispatcher.queue", attributes: .concurrent)
    
    func publishGreenMessage(_ msg: GreenMessage) {
        queue.async(flags: .barrier) {
            self.greenSubscribers.forEach { $0.value(msg) }
        }
    }
    
    func publishBlueMessage(_ msg: BlueMessage) {
        queue.async(flags: .barrier) {
            self.blueSubscribers.forEach { $0.value(msg) }
        }
    }
    
    func publishOrangeMessage(_ msg: OrangeMessage) {
        queue.async(flags: .barrier) {
            self.orangeSubscribers.forEach { $0.value(msg) }
        }
    }
    
    @discardableResult
    func subscribeToGreenMessage(_ callback: @escaping (GreenMessage) -> Void) -> UUID {
        let id = UUID()
        queue.async(flags: .barrier) {
            self.greenSubscribers[id] = callback
        }
        return id
    }
    
    @discardableResult
    func subscribeToBlueMessage(_ callback: @escaping (BlueMessage) -> Void) -> UUID {
        let id = UUID()
        queue.async(flags: .barrier) {
            self.blueSubscribers[id] = callback
        }
        return id
    }
    
    @discardableResult
    func subscribeToOrangeMessage(_ callback: @escaping (OrangeMessage) -> Void) -> UUID {
        let id = UUID()
        queue.async(flags: .barrier) {
            self.orangeSubscribers[id] = callback
        }
        return id
    }
    
    func unsubscribeFromGreenMessage(_ id: UUID) {
        queue.async(flags: .barrier) {
            self.greenSubscribers.removeValue(forKey: id)
        }
    }
    
    func unsubscribeFromBlueMessage(_ id: UUID) {
        queue.async(flags: .barrier) {
            self.blueSubscribers.removeValue(forKey: id)
        }
    }
    
    func unsubscribeFromOrangeMessage(_ id: UUID) {
        queue.async(flags: .barrier) {
            self.orangeSubscribers.removeValue(forKey: id)
        }
    }
}

func generateGreenMessages(dispatcher: MessageDispatcher, group: DispatchGroup) {
    group.enter()
    defer { group.leave() }
    for i in 0..<10 {
        let msg = GreenMessage(text: "GreenMessage_\(i)", counter: Int.random(in: 1...100))
        dispatcher.publishGreenMessage(msg)
        Thread.sleep(forTimeInterval: 0.1)
    }
}

func generateBlueMessages(dispatcher: MessageDispatcher, group: DispatchGroup) {
    group.enter()
    defer { group.leave() }
    for i in 0..<10 {
        let msg = BlueMessage(value1: Double.random(in: 0.0...100.0), value2: Double.random(in: 0.0...100.0))
        dispatcher.publishBlueMessage(msg)
        Thread.sleep(forTimeInterval: 0.15)
    }
}

func generateOrangeMessages(dispatcher: MessageDispatcher, group: DispatchGroup) {
    group.enter()
    defer { group.leave() }
    for i in 0..<10 {
        let msg = OrangeMessage(
            text1: "OrangeMessage1_\(i)",
            text2: "OrangeMessage2_\(i)",
            integer: Int.random(in: 1...100),
            value: Double.random(in: 0.0...100.0)
        )
        dispatcher.publishOrangeMessage(msg)
        Thread.sleep(forTimeInterval: 0.2)
    }
}

let dispatcher = MessageDispatcher()

let greenSubscriptionId = dispatcher.subscribeToGreenMessage { msg in
    print("Received GreenMessage: \(msg.text), \(msg.counter)")
}
let blueSubscriptionId = dispatcher.subscribeToBlueMessage { msg in
    print("Received BlueMessage: \(msg.value1), \(msg.value2)")
}
let orangeSubscriptionId = dispatcher.subscribeToOrangeMessage { msg in
    print("Received OrangeMessage: \(msg.text1), \(msg.text2), \(msg.integer), \(msg.value)")
}

DispatchQueue.global().asyncAfter(deadline: .now() + 1) {
    dispatcher.unsubscribeFromGreenMessage(greenSubscriptionId)
}

let group = DispatchGroup()

DispatchQueue.global().async(group: group) {
    generateGreenMessages(dispatcher: dispatcher, group: group)
}
DispatchQueue.global().async(group: group) {
    generateBlueMessages(dispatcher: dispatcher, group: group)
}
DispatchQueue.global().async(group: group) {
    generateOrangeMessages(dispatcher: dispatcher, group: group)
}

group.wait()
print("All messages generated.")
